### PR TITLE
parser: support two json grammar sugar.

### PR DESCRIPTION
**Pull request by hicqu**
Support 2 json grammar sugar:
1. `select a->'$.a' from t` equals to `select json_extract(a, '$.a') from t`;
2. `select a->>'$.a' from t` equals to `select json_unquote(json_extract(a, '$.a')) from t`.

### CODE REVIEW

**Comment by hicqu** [comment]
@winkyao @tiancaiamao @hanfei1991 , PTAL, thanks.

File: parser/lexer.go [hunk 1]
```diff
@@ -294,8 +294,18 @@ func startWithSharp(s *Scanner) (tok int, pos Pos, lit string) {
 func startWithDash(s *Scanner) (tok int, pos Pos, lit string) {
 	pos = s.r.pos()
 	if !strings.HasPrefix(s.r.s[pos.Offset:], "-- ") {
-		tok = int('-')
-		s.r.inc()
+		if strings.HasPrefix(s.r.s[pos.Offset:], "->") {
```
Comment (tiancaiamao) [inline parser/lexer.go]: `if` nested too much, you'd better reverse the logic:

```
if strings.HasPrefix(s.r.s[pos.Offset:], "-- ") {
    // skip comment and read next token
    // return
}
if strings.HasPrefix(s.r.s[pos.Offset:], "->") {
    // this branch test -> and ->>
    // return
}
tok = int('-')
s.r.inc()
return
```

Special cases goes first, and returns as soon as possible, followed by normal case.
This way the code can be cleaner.

File: parser/misc.go [hunk 2]
```diff
@@ -119,6 +119,8 @@ func init() {
 	initTokenString("<>", neqSynonym)
 	initTokenString("<<", lsh)
 	initTokenString(">>", rsh)
+	initTokenString("->", jss)   // jss means json separator symbol.
```
Comment (tiancaiamao) [inline parser/misc.go]: Those two line is useless, as the exists of `initTokenFunc("-", startWithDash)`, remove them.

File: parser/parser.y [hunk 3]
```diff
@@ -560,6 +560,8 @@ import (
 	nulleq		"<=>"
 	placeholder	"PLACEHOLDER"
 	rsh		">>"
+	jss		"->"
```
Comment (tiancaiamao) [inline parser/parser.y]: Format it, sort by ascii order.

File: parser/parser_test.go [hunk 4]
```diff
@@ -1052,6 +1052,14 @@ func (s *testParserSuite) TestBuiltin(c *C) {
...
 		{`SELECT JSON_TYPE('[123]');`, true},
 		{`SELECT JSON_TYPE();`, true},
+
+		// For two json grammar sugar.
+		{`SELECT a->'$.a' FROM t`, true},
+		{`SELECT a->>'$.a' FROM t`, true},
+		{`SELECT '{}'->'$.a' FROM t`, false},
+		{`SELECT '{}'->>'$.a' FROM t`, false},
+		{`SELECT a->3 FROM t`, false},
+		{`SELECT a->>3 FROM t`, false},
```
Comment (tiancaiamao) [inline parser/parser_test.go]: Please add some tests to executor too.

File: parser/parser.y [hunk 5]
```diff
@@ -2761,6 +2763,29 @@ Function:
...
+|	Identifier jss stringLit
+	{
+	    col := &ast.ColumnNameExpr{Name: &ast.ColumnName{Name: model.NewCIStr($1)}}
+
+	    tp := types.NewFieldType(mysql.TypeString)
+	    tp.Charset, tp.Collate = parser.charset, parser.collation
+	    expr := ast.NewValueExpr($3)
+	    expr.SetType(tp)
+
+	    $$ = &ast.FuncCallExpr{FnName: model.NewCIStr("JSON_EXTRACT"), Args: []ast.ExprNode{col, expr}}
```
Comment (shenli) [inline parser/parser.y]: Use the const defined in ast package for "JSON_EXTRACT".

File: parser/parser.y [hunk 6]
```diff
@@ -2761,6 +2763,29 @@ Function:
...
+|	Identifier juss stringLit
+	{
+	    col := &ast.ColumnNameExpr{Name: &ast.ColumnName{Name: model.NewCIStr($1)}}
+
+	    tp := types.NewFieldType(mysql.TypeString)
+	    tp.Charset, tp.Collate = parser.charset, parser.collation
+	    expr := ast.NewValueExpr($3)
+	    expr.SetType(tp)
+
+	    extract := &ast.FuncCallExpr{FnName: model.NewCIStr("JSON_EXTRACT"), Args: []ast.ExprNode{col, expr}}
```
Comment (shenli) [inline parser/parser.y]: ditto

**Comment by tiancaiamao** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**